_posts/2021-12-3-arrow-duckdb.md

```diff
@@ -0,0 +1,383 @@
+arrow::copy_files("s3://ursa-labs-taxi-data", "nyc-taxi")
+```
+
+
+#### Python
+```py
+# Reads dataset partitioning it in year/month folder
+nyc_dataset = ds.dataset('nyc-taxi/', partitioning=["year", "month"])
+
+# Gets Database Connection
+con = duckdb.connect()
+
+query = con.execute("SELECT * FROM nyc")
+# DuckDB's queries can now produce a Record Batch Reader
+record_batch_reader = query.fetch_record_batch()
+# Which means we can stream the whole query per batch.
+# This retrieves the first batch
+chunk = record_batch_reader.read_next_batch()
+```
+#### R
+```r
+# Reads dataset partitioning it in year/month folder
+nyc_dataset = open_dataset("nyc-taxi/", partitioning = c("year", "month"))
+
+# Gets Database Connection
+con <- dbConnect(duckdb::duckdb())
+
+# We can use the same function as before to register our arrow dataset
+duckdb::duckdb_register_arrow(con, "nyc", nyc_dataset)
+
+res <- dbSendQuery(con, "SELECT * FROM nyc", arrow = TRUE)
+# DuckDB's queries can now produce a Record Batch Reader
+record_batch_reader <- duckdb::duckdb_fetch_record_batch(res)
+
+# Which means we can stream the whole query per batch.
+# This retrieves the first batch
+cur_batch <- record_batch_reader$read_next_batch()
+```
+
+The preceding R code shows in low-level detail how the data is streaming. We provide the helper `to_arrow()` in the Arrow package which is a wrapper around this that makes it easy to incorporate this streaming into a dplyr pipeline. [^1]
+
+## Benchmark Comparison
+
+Here we demonstrate in a simple benchmark the performance difference between querying Arrow datasets with DuckDB and querying Arrow datasets with Pandas.
+For both the Projection and Filter pushdown comparison, we will use Arrow tables. That is due to Pandas not being capable of consuming Arrow stream objects.
+
+### Projection Pushdown
+
+In this example we run a simple aggregation on two columns of our lineitem table.
+
+``` python
+# DuckDB
+lineitem = pq.read_table('lineitemsf1.snappy.parquet')
+con = duckdb.connect()
+
+# Transforms Query Result from DuckDB to Arrow Table
+con.execute("""SELECT sum(l_extendedprice * l_discount) AS revenue
+                FROM
+                lineitem;""").fetch_arrow_table()
+
+```
+
+``` python
+# Pandas
+arrow_table = pq.read_table('lineitemsf1.snappy.parquet')
+
+# Converts an Arrow table to a Dataframe
+df = arrow_table.to_pandas()
+
+# Runs aggregation
+res =  pd.DataFrame({'sum': [(df.l_extendedprice * df.l_discount).sum()]})
+
+# Creates an Arrow Table from a Dataframe
+new_table = pa.Table.from_pandas(res)
+
+```
+
+|    Name     | Time (s) |
+|-------------|---------:|
+| DuckDB  | 0.19    |
+| Pandas      | 2.13    |
+
+The lineitem table is composed of 16 columns, however, to execute this query only two columns ```l_extendedprice ``` and  *  ``` l_discount ```are necessary. Since DuckDB can push down the projection of these columns, it is capable of executing this query about one order of magnitude faster than Pandas.
```

Comment:
```suggestion
The lineitem table is composed of 16 columns, however, to execute this query only two columns ```l_extendedprice``` and  *  ```l_discount``` are necessary. Since DuckDB can push down the projection of these columns, it is capable of executing this query about one order of magnitude faster than Pandas.
```

Very minor spacing adjustments @pdet are these meaningful?

Reply:
I think that looks fine.